src/server/checker/CheckerService.ts

```diff
@@ -164,6 +190,114 @@ export class CheckerService {
     return checker
   }
 
+  private findAndCheckAuthOwner: (
+    id: string,
+    user: User | null
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ) => Promise<(Model<Checker, any> & Checker) | null> = async (id, user) => {
+    const checker = await this.CheckerModel.findByPk(id, {
+      include: [this.UserModel],
+    })
+    if (checker) {
+      const isAuthorized = checker.users?.some(
+        // Sequelize automatically adds properties of userToChecker association to user model
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        (u) => u.UserToChecker.isOwner && u.id === user?.id
+      )
+      if (!isAuthorized) {
+        throw new Error('Unauthorized')
+      }
+    }
+    return checker
+  }
+
+  listCollaborators: (id: string, user: User) => Promise<CollaboratorUser[]> =
+    async (id, user) => {
+      const checker = await this.findAndCheckAuth(id, user)
+      if (checker) {
+        // Sequelize automatically adds users property to checker model through association
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        return checker.users
+      } else return []
+    }
+
+  addCollaborator: (
+    id: string,
+    user: User,
+    collaboratorEmail: string
+  ) => Promise<void> = async (id, user, collaboratorEmail) => {
+    const transaction = await this.sequelize.transaction()
+    const options = this.isSqliteFile ? {} : { transaction }
+
+    try {
+      const checker = await this.findAndCheckAuth(id, user)
+      collaboratorEmail = collaboratorEmail.toLowerCase()
+
+      if (checker) {
+        const collaboratorUser = await this.UserModel.findOne({
+          where: { email: collaboratorEmail },
+          ...options,
+        })
+        // Sequelize automatically adds users property to checker model through association
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        if (checker.users?.some((u) => u.email === collaboratorEmail))
```

Comment:
If we are using the `checker` returned by `findAndCheckAuth`, then it should also be included in the same transaction. Same for the other methods.